src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+                        rangeFacts |= RangeFacts.IsGenerator;
+                        break;
+                    case SyntaxKind.ReturnStatement:
+                        if (permittedJumps & PermittedJumps.Return) {
+                            rangeFacts |= RangeFacts.HasReturn;
+                        }
+                        else {
+                            (errors || (errors = [])).push(createDiagnosticForNode(node, Messages.CannotExtractRangeContainingConditionalReturnStatement));
+                        }
+                        break;
+                    default:
+                        forEachChild(node, visit);
+                        break;
+                }
+
+                permittedJumps = savedPermittedJumps;
+            }
+        }
+    }
+
+    function isValidExtractionTarget(node: Node) {
+        // Note that we don't use isFunctionLike because we don't want to put the extracted closure *inside* a method
+        return (node.kind === SyntaxKind.FunctionDeclaration) || (node.kind === SyntaxKind.Constructor) || isSourceFile(node) || isModuleBlock(node) || isClassLike(node);
+    }
+
+    /**
+     * Computes possible places we could extract the function into. For example,
+     * you may be able to extract into a class method *or* local closure *or* namespace function,
+     * depending on what's in the extracted body.
+     */
+    export function collectEnclosingScopes(range: TargetRange): Scope[] | undefined {
+        let current: Node = isArray(range.range) ? firstOrUndefined(range.range) : range.range;
+        if (range.facts & RangeFacts.UsesThis) {
+            // if range uses this as keyword or as type inside the class then it can only be extracted to a method of the containing class
+            const containingClass = getContainingClass(current);
+            if (containingClass) {
+                return [containingClass];
+            }
+        }
+
+        const start = current;
+
+        let scopes: Scope[] | undefined = undefined;
+        while (current) {
+            // We want to find the nearest parent where we can place an "equivalent" sibling to the node we're extracting out of.
+            // Walk up to the closest parent of a place where we can logically put a sibling:
+            //  * Function declaration
+            //  * Class declaration or expression
+            //  * Module/namespace or source file
+            if (current !== start && isValidExtractionTarget(current)) {
+                (scopes = scopes || []).push(current as FunctionLikeDeclaration);
+            }
+
+            // A function parameter's initializer is actually in the outer scope, not the function declaration
+            if (current && current.parent && current.parent.kind === SyntaxKind.Parameter) {
+                // Skip all the way to the outer scope of the function that declared this parameter
+                current = current.parent.parent.parent;

```

Comment:
>current.parent.parent.parent [](start = 26, length = 28)

This makes me a bit nervous.  Consider asserting that the result is of an appropriate kind.